crates/oxc_linter/src/rules/constructor_super.rs

```diff
@@ -0,0 +1,131 @@
+struct ConstructorSuperDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(constructor-super): Unexpected 'super()' because 'super' is not a constructor.")]
+#[diagnostic(severity(warning), help("Do not call 'super()' from constructor."))]
+struct SuperNotConstructorDiagnostic(
+    #[label("unexpected 'super()'")] pub Span,
+    #[label("because this is not a constructor")] pub Span,
+);
+
+#[derive(Debug, Default, Clone)]
+pub struct ConstructorSuper;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Require 'super()' calls in constructors.
+    ///
+    /// ### Why is this bad?
+    ///
+    ///
+    /// ### Example
+    /// ```javascript
+    /// class A extends B {
+    ///   constructor() {}
+    /// }
+    /// ```
+    ConstructorSuper
+);
+
+impl Rule for ConstructorSuper {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::Class(class) = node.get().kind() else { return };
+        let Some(ctor) = class.body.body.iter().find_map(|el| match el {
+            ClassElement::MethodDefinition(method_definition)
+                if method_definition.kind == MethodDefinitionKind::Constructor =>
+            {
+                Some(method_definition)
+            }
+            _ => None,
+        }) else { return };
+
+        // In cases where there's no super-class, calling 'super()' inside the constructor
+        // is handled by the parser.
+        if let Some(super_class) = &class.super_class {
+            ctor.value.body.as_ref().map_or_else(|| {
+                ctx.diagnostic(ConstructorSuperDiagnostic(ctor.span));
+            }, |function_body| {
+                let super_call_expr = function_body.statements.iter().find_map(|stmt| {
+                    let Statement::ExpressionStatement(expr) = stmt else { return None };
+                    let Expression::CallExpression(call_expr) = &expr.expression else { return None };
+                    if matches!(call_expr.callee, Expression::Super(_)) {
+                        Some(call_expr.span)
+                    } else {
+                        None
+                    }
+                });
+
+                super_call_expr.map_or_else(|| {
+                    ctx.diagnostic(ConstructorSuperDiagnostic(ctor.span));
+                }, |span| {
+                    if let Some(super_class_span) = super_class.span() {
+                        ctx.diagnostic(SuperNotConstructorDiagnostic(span, super_class_span));
+                    }
+                });
+            });
+        }
+    }
+}
+
+trait NonConstructor {
+    fn span(&self) -> Option<Span>;
+}
+
+impl<'a> NonConstructor for Expression<'a> {
+    fn span(&self) -> Option<Span> {
+        match self {
+            Self::NullLiteral(lit) => Some(lit.span),
+            Self::NumberLiteral(lit) => Some(lit.span),
+            Self::StringLiteral(lit) => Some(lit.span),
```

Comment:
There are more things to handle that are considered a "non-constructor" but I haven't yet covered them all